### ISSUE TITLE
NullPointerException when trying to update a Redshift Database in Metabase UI

### ISSUE THREAD

**Issue by taufort** [metabase/metabase#32787]
**Describe the bug**

We had to upgrade from Metabase `0.45.2` to Metabase `0.46.6.4` to bypass the [Google sign-in issue](https://github.com/metabase/metabase/issues/32602) because none of our users in production could log in anymore. The upgrade fixed the Google sign-in issue but a new bug appeared specifically with our Redshift databases.

We noticed that when trying to access a dashboard querying Redshift databases, the queries failed with the following message in the UI: `Connections could not be acquired from the underlying database!` (see screenshot attached below). Following the troubleshooting guides of Metabase, we tried to add the JDBC option `trustServerCertificate=true` in the database configuration and that's when we found out we could not update the Redshift database anymore (see screenshot attached below) because of a 400 HTTP error.

**Logs**

Here are the logs when we try to update a Redshift Database:
```
2023-08-01 10:03:30,812 DEBUG middleware.log :: GET /api/database 200 14.3 ms (4 DB calls) App DB connections: 0/15 Jetty threads: 3/50 (7 idle, 0 queued) (62 total active threads) Queries in flight: 0 (0 queued)
2023-08-01 10:03:31,999 DEBUG middleware.log :: GET /api/database/6 200 13.7 ms (3 DB calls) App DB connections: 0/15 Jetty threads: 3/50 (7 idle, 0 queued) (62 total active threads) Queries in flight: 0 (0 queued)
2023-08-01 10:03:34,247 ERROR driver.util :: Failed to connect to Database
java.lang.NullPointerException
	at clojure.lang.Reflector.invokeInstanceMethod(Reflector.java:97)
	at metabase.driver.connection$decorate_and_fix$fn__14957.invoke(connection.clj:40)
	at metabase.driver.connection.proxy$java.lang.Object$Connection$1d5212e.close(Unknown Source)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:302)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:299)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_$fn__50177.invoke(connection.clj:311)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection$fn__50159.invoke(connection.clj:287)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:141)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:130)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invokeStatic(connection.clj:285)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invoke(connection.clj:281)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:310)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:306)
	at metabase.driver.sql_jdbc$fn__105199.invokeStatic(sql_jdbc.clj:43)
	at metabase.driver.sql_jdbc$fn__105199.invoke(sql_jdbc.clj:41)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__46887.invoke(util.clj:144)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2023-08-01 10:03:34,253 ERROR api.database :: Cannot connect to Database
java.lang.NullPointerException
	at clojure.lang.Reflector.invokeInstanceMethod(Reflector.java:97)
	at metabase.driver.connection$decorate_and_fix$fn__14957.invoke(connection.clj:40)
	at metabase.driver.connection.proxy$java.lang.Object$Connection$1d5212e.close(Unknown Source)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:302)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:299)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_$fn__50177.invoke(connection.clj:311)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection$fn__50159.invoke(connection.clj:287)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:141)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:130)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invokeStatic(connection.clj:285)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invoke(connection.clj:281)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:310)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:306)
	at metabase.driver.sql_jdbc$fn__105199.invokeStatic(sql_jdbc.clj:43)
	at metabase.driver.sql_jdbc$fn__105199.invoke(sql_jdbc.clj:41)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__46887.invoke(util.clj:144)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2023-08-01 10:03:34,255 DEBUG middleware.log :: PUT /api/database/6 400 22.8 ms (3 DB calls)
{:message nil}
```

Here are the Javascript logs in the Developer console: 
[javascript_console.log](https://github.com/metabase/metabase/files/12227363/javascript_console.log)

**To Reproduce**

Unfortunately, we were not able to reproduce this NPE on another Metabase application for now besides our production application.

**Expected behavior**

We should be able to update the Redshift Database configuration without any 400 HTTP error.

**Screenshots**

<img width="673" alt="connection_error" src="https://github.com/metabase/metabase/assets/26935026/66bfa68c-0971-4d59-8668-e4959f9c38f0">

<img width="656" alt="cannot_update_redshift_database" src="https://github.com/metabase/metabase/assets/26935026/b7d3a5f5-4a85-48d5-8fe6-91ba73f310b7">

**Severity**

The majority of our dashboards are based on Redshift databases. We only have a few dashboards based on Databricks databases still available. As a result, this issue is quite severe for our business as most users cannot use Redshift dashboards anymore.

**Additional context**

We've tried to update the RedShift Database details directly in the `public.metabase_database` table in our Metabase PostgreSQL database to add the `trustServerCertificate=true` option but unfortunately, it did not change anything and the NPE issue is still there.

Here is the Database details column content when we first saw the bug:
```
{"additional-options":null,"db":"production","password":"****","port":5439,"advanced-options":false,"schema-filters-type":"all","host":"****","tunnel-enabled":false,"user":"****"}
```
Here is the new content we set trying to bypass the NPE bug with no effect whatsoever:
```
{"additional-options":"trustServerCertificate=true","ssl":true,"db":"production","password":"****","port":5439,"advanced-options":true,"schema-filters-type":"all","host":"****","tunnel-enabled":false,"user":"****"}
```

**Metabase Diagnostic Info**

```json
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.20+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.20",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.20+8",
    "os.name": "Linux",
    "os.version": "5.10.173-154.642.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "sparksql-databricks",
      "redshift"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "12.14"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.5.1"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2023-07-28",
      "tag": "v0.46.6.4",
      "branch": "release-x.46.6.x",
      "hash": "7c60aca"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Comment by taufort** [comment]
We identified that the bug does not happen when we don't have the Spark Databricks Metabase driver in the classpath so it would look like it's a driver conflict bug. I have created a [new issue on the Spark Databricks driver GitHub repository](https://github.com/relferreira/metabase-sparksql-databricks-driver/issues/23).

**Comment by paoliniluis** [comment]
Thanks, closing this issue then

**Comment by metamben** [comment]
This problem surfaces with other drivers too. See https://metaboat.slack.com/archives/C0439QDFHFZ/p1708697829820119

**Comment by metamben** [comment]
Had a look at [metabase-sparksql-databricks-driver](https://github.com/relferreira/metabase-sparksql-databricks-driver/tree/5e1c3b5507853f533d03789c9c6e2b49c2a2f838) and I think the problem is that it doesn't handle the case when the wrapped databricks driver returns `null`. Instead of passing that through (signifying that the databricks driver cannot connect to redshift, clickhouse, etc.), it wraps it and returns an object. The JDBC DriverManager then assumes that the connection was successful and doesn't even try connecting to the right database.

Long story short, `metabase-sparksql-databricks-driver` needs fixing.
